{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import rdflib\n",
    "import re\n",
    "import pandas as pd\n",
    "from rdflib import URIRef, BNode, Literal, Graph, RDF, Namespace, ODRL2, RDFS\n",
    "import requests\n",
    "import time\n",
    "from tqdm import tqdm\n",
    "import urllib.parse\n",
    "from maayanlab_bioinformatics.utils import fetch_save_read, merge\n",
    "from functools import lru_cache\n",
    "import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(('NOS1', '4842'), (None, 'None'))"
      ]
     },
     "execution_count": 179,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "@lru_cache()\n",
    "def ncbi_genes_fetch(organism='Mammalia/Homo_sapiens', filters=lambda ncbi: (ncbi['type_of_gene'] == \"protein-coding\")):\n",
    "    ''' Fetch the current NCBI Human Gene Info database.\n",
    "    See ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/ for the directory/file of the organism of interest.\n",
    "    '''\n",
    "    def maybe_split(record):\n",
    "        ''' NCBI Stores Nulls as '-' and lists '|' delimited\n",
    "        '''\n",
    "        if record in {'', '-'}:\n",
    "            return set()\n",
    "        return set(record.split('|'))\n",
    "    #\n",
    "    def supplement_dbXref_prefix_omitted(ids):\n",
    "        ''' NCBI Stores external IDS with Foreign:ID while most datasets just use the ID\n",
    "        '''\n",
    "        for id in ids:\n",
    "          # add original id\n",
    "          yield id\n",
    "          # also add id *without* prefix\n",
    "          if ':' in id:\n",
    "            yield id.split(':', maxsplit=1)[1]\n",
    "    #\n",
    "    ncbi = fetch_save_read(\n",
    "    'ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/{}.gene_info.gz'.format(organism),\n",
    "    '{}.gene_info.tsv'.format(organism),\n",
    "    sep='\\t',\n",
    "    )\n",
    "    if filters and callable(filters):\n",
    "        ncbi = ncbi[filters(ncbi)]\n",
    "    #\n",
    "    ncbi['All_synonyms'] = [\n",
    "        set.union(\n",
    "          maybe_split(gene_info['Symbol']),\n",
    "          maybe_split(gene_info['Symbol'].upper()),\n",
    "          maybe_split(gene_info['Symbol_from_nomenclature_authority']),\n",
    "          maybe_split(str(gene_info['GeneID'])),\n",
    "          maybe_split(gene_info['Synonyms']),\n",
    "          maybe_split(gene_info['Other_designations']),\n",
    "          maybe_split(gene_info['LocusTag']),\n",
    "          set(supplement_dbXref_prefix_omitted(maybe_split(gene_info['dbXrefs']))),\n",
    "        )\n",
    "        for _, gene_info in ncbi.iterrows()\n",
    "    ]\n",
    "    return ncbi\n",
    "\n",
    "@lru_cache()                  \n",
    "def ncbi_genes_lookup(organism='Mammalia/Homo_sapiens', filters=lambda ncbi: ncbi['type_of_gene']=='protein-coding'):\n",
    "    ''' Return a lookup dictionary with synonyms as the keys, and official symbols as the values\n",
    "    Usage:\n",
    "    ```python\n",
    "    ncbi_lookup = ncbi_genes_lookup('Mammalia/Homo_sapiens')\n",
    "    print(ncbi_lookup('STAT3')) # any alias will get converted into the official symbol\n",
    "    ```\n",
    "    '''\n",
    "    ncbi_genes = ncbi_genes_fetch(organism=organism)\n",
    "    synonyms, symbols, gene_ids = zip(*{\n",
    "    (synonym, gene_info['Symbol'], gene_info['GeneID'])\n",
    "    for _, gene_info in ncbi_genes.iterrows()\n",
    "    for synonym in gene_info['All_synonyms']\n",
    "    })\n",
    "    ncbi_lookup = pd.Series(symbols, index=synonyms)\n",
    "    index_values = ncbi_lookup.index.value_counts()\n",
    "    ncbi_lookup_disambiguated = ncbi_lookup.drop(index_values[index_values > 1].index)\n",
    "    for i in symbols:\n",
    "        if i not in ncbi_lookup_disambiguated.index:\n",
    "            ncbi_lookup_disambiguated[i] = i\n",
    "    \n",
    "    ncbi_lookup_id = pd.Series(gene_ids, index=synonyms)\n",
    "    index_values_id = ncbi_lookup_id.index.value_counts()\n",
    "    ncbi_lookup_disambiguated_id = ncbi_lookup_id.drop(index_values_id[index_values_id > 1].index)\n",
    "    ncbi_genes = ncbi_genes.set_index(\"GeneID\")\n",
    "    for i in gene_ids:\n",
    "        gene = ncbi_genes.at[i, \"Symbol\"]\n",
    "        if gene not in ncbi_lookup_disambiguated_id.index:\n",
    "            ncbi_lookup_disambiguated_id[gene] = i\n",
    "\n",
    "    return ncbi_lookup_disambiguated.to_dict().get, ncbi_lookup_disambiguated_id.get\n",
    "\n",
    "symbol_mapper, gene_id_mapper = ncbi_genes_lookup()\n",
    "def get_info(gene):\n",
    "    return(symbol_mapper(gene), str(gene_id_mapper(gene)))\n",
    "\n",
    "get_info('NOS1'), get_info('HAUS1P1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "label_id_mapper = {\n",
    "    \"LY-294002\": \"CID:3973\",\n",
    "    \"PD-98059\": \"CID:4713\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pLI</th>\n",
       "      <th>Residual.Variation.Intolerance.Score</th>\n",
       "      <th>Residual.Variation.Intolerance.Score.Percentile</th>\n",
       "      <th>pHI</th>\n",
       "      <th>pTS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>9.064924e-05</td>\n",
       "      <td>-0.466531</td>\n",
       "      <td>23.513800</td>\n",
       "      <td>0.13450</td>\n",
       "      <td>0.52075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>3.619701e-03</td>\n",
       "      <td>-0.378346</td>\n",
       "      <td>28.013682</td>\n",
       "      <td>0.21197</td>\n",
       "      <td>0.22586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>5.401149e-04</td>\n",
       "      <td>0.099179</td>\n",
       "      <td>60.757254</td>\n",
       "      <td>0.92010</td>\n",
       "      <td>0.39196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>1.329022e-22</td>\n",
       "      <td>2.419901</td>\n",
       "      <td>98.525596</td>\n",
       "      <td>0.27137</td>\n",
       "      <td>0.18193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A4GALT</th>\n",
       "      <td>1.115987e-01</td>\n",
       "      <td>-0.290161</td>\n",
       "      <td>33.339231</td>\n",
       "      <td>0.14511</td>\n",
       "      <td>0.09411</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                      pLI  Residual.Variation.Intolerance.Score  \\\n",
       "Gene Symbol                                                       \n",
       "A1BG         9.064924e-05                             -0.466531   \n",
       "A1CF         3.619701e-03                             -0.378346   \n",
       "A2M          5.401149e-04                              0.099179   \n",
       "A2ML1        1.329022e-22                              2.419901   \n",
       "A4GALT       1.115987e-01                             -0.290161   \n",
       "\n",
       "             Residual.Variation.Intolerance.Score.Percentile      pHI      pTS  \n",
       "Gene Symbol                                                                     \n",
       "A1BG                                               23.513800  0.13450  0.52075  \n",
       "A1CF                                               28.013682  0.21197  0.22586  \n",
       "A2M                                                60.757254  0.92010  0.39196  \n",
       "A2ML1                                              98.525596  0.27137  0.18193  \n",
       "A4GALT                                             33.339231  0.14511  0.09411  "
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_scores = pd.read_csv(\"https://raw.githubusercontent.com/nih-cfde/ReproToxTables/main/Susceptibility%20Scores%20and%20GWAS%20Gene%20Lists/Susceptibility%20Scores/All%20Scores.csv\", index_col=0)\n",
    "gene_scores.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Similarity_Score</th>\n",
       "      <th>Rank</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FTI-276</th>\n",
       "      <td>0.111335</td>\n",
       "      <td>33473.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRD-K08703257</th>\n",
       "      <td>0.104131</td>\n",
       "      <td>33472.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>temozolomide</th>\n",
       "      <td>0.102109</td>\n",
       "      <td>33471.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Gossypetin</th>\n",
       "      <td>0.101613</td>\n",
       "      <td>33470.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAK-715</th>\n",
       "      <td>0.100779</td>\n",
       "      <td>33469.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Similarity_Score     Rank\n",
       "FTI-276                0.111335  33473.0\n",
       "BRD-K08703257          0.104131  33472.0\n",
       "temozolomide           0.102109  33471.0\n",
       "Gossypetin             0.101613  33470.0\n",
       "TAK-715                0.100779  33469.0"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_scores = pd.read_csv(\"data/L1000_2021_Signature_Similarity_predicted_drug_table.csv\", index_col=0)\n",
    "drug_scores[\"Rank\"] = drug_scores.Similarity_Score.rank(method=\"first\")\n",
    "drug_scores.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HPO description</th>\n",
       "      <th>KF Frequency</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HPO ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HP:0410030</th>\n",
       "      <td>Cleft lip</td>\n",
       "      <td>888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0000776</th>\n",
       "      <td>Congenital diaphragmatic hernia</td>\n",
       "      <td>770</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0000175</th>\n",
       "      <td>Cleft palate</td>\n",
       "      <td>704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0010438</th>\n",
       "      <td>Abnormal ventricular septum morphology</td>\n",
       "      <td>629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HP:0001710</th>\n",
       "      <td>Conotruncal defect</td>\n",
       "      <td>545</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   HPO description  KF Frequency\n",
       "HPO ID                                                          \n",
       "HP:0410030                               Cleft lip           888\n",
       "HP:0000776         Congenital diaphragmatic hernia           770\n",
       "HP:0000175                            Cleft palate           704\n",
       "HP:0010438  Abnormal ventricular septum morphology           629\n",
       "HP:0001710                      Conotruncal defect           545"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "HPO_Freq = pd.read_csv(\"data/HPO_Freq.tsv\", sep=\"\\t\", index_col=1)\n",
    "HPO_Freq.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "associations = {\n",
    "  'https://raw.githubusercontent.com/nih-cfde/ReproToxTables/main/CDC-birth-defects/Geneshot_BirthDefects_Gene_Associations.ttl': \"Geneshot\",\n",
    "  'https://raw.githubusercontent.com/nih-cfde/ReproToxTables/main/CDC-birth-defects/Drugshot_BirthDefects_Drug_Associations.ttl': \"Drugshot\",\n",
    "  'https://raw.githubusercontent.com/nih-cfde/ReproToxTables/main/CDC-birth-defects/DrugEnrichr_BirthDefects_Drug_Associations.ttl': \"DrugEnrichr\",\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"data/birth_defects.json\") as o:\n",
    "    birth_defects = json.loads(o.read())\n",
    "\n",
    "with open(\"data/drug_ref.json\") as o:\n",
    "    drugs = json.loads(o.read())\n",
    "    for k,v in drugs.items():\n",
    "        new_dict = {}\n",
    "        for i,j in v.items():\n",
    "            new_dict[i.replace(\".\", \"_\")] = j\n",
    "        drugs[k] = new_dict\n",
    "\n",
    "with open(\"data/gene_ref.json\") as o:\n",
    "    genes = json.loads(o.read())\n",
    "    for k,v in genes.items():\n",
    "        new_dict = {}\n",
    "        for i,j in v.items():\n",
    "            new_dict[i.replace(\".\", \"_\")] = j\n",
    "        genes[k] = new_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "for k,v in genes.items():\n",
    "    a = v.pop(\"Residual.Variation.Intolerance.Score\", None)\n",
    "    if a:\n",
    "        v[\"Residual_Variation_Intolerance_Score\"] = a\n",
    "    a = v.pop(\"Residual.Variation.Intolerance.Score.Percentile\", None)\n",
    "    if a:\n",
    "        v[\"Residual_Variation_Intolerance_Score_Percentile\"] = a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "iri_mappers = [\n",
    "  dict(\n",
    "    cls='BirthDefect',\n",
    "    expr=re.compile(r'^https://www\\.orpha\\.net/ORDO/(?P<id>.+)$'),\n",
    "  ),\n",
    "  dict(\n",
    "    cls='BirthDefect',\n",
    "    expr=re.compile(r'^https://purl\\.obolibrary\\.org/obo/(?P<id>.+)$'),\n",
    "  ),\n",
    "  dict(\n",
    "    cls='Gene',\n",
    "    expr=re.compile(r'^https://identifiers\\.org/hgnc\\.symbol/(?P<symbol>.+)$'),\n",
    "  ),\n",
    "  dict(\n",
    "    cls='Drug',\n",
    "    expr=re.compile(r'^https://identifiers\\.org/lincs\\.smallmolecule:(?P<id>.+)$'),\n",
    "  ),\n",
    "  # dict(\n",
    "  #   cls='Relationship',\n",
    "  #   expr=re.compile(r'^https://semanticscience\\.org/resource/(?P<id>.+)$'),\n",
    "  # ),\n",
    "  dict(\n",
    "    cls='BirthDefect',\n",
    "    expr=re.compile(r'^https://hpo\\.jax\\.org/app/browse/term/(?P<id>.+)$'),\n",
    "  ),\n",
    "]\n",
    "\n",
    "relationships = {\n",
    "  \"SIO_010299\": \"disease\",\n",
    "  \"SIO_010056\": \"phenotype\",\n",
    "  \"SIO_000993\": \"chemical-disease\",\n",
    "  \"SIO_000983\": \"gene-disease\"\n",
    "}\n",
    "\n",
    "orpha = {}\n",
    "def iri_to_node(iri):\n",
    "    for mapper in iri_mappers:\n",
    "        iri = iri.replace(\"HP_\", \"HP:\")\n",
    "        m = mapper['expr'].match(iri)\n",
    "        if m:\n",
    "            group_dict = m.groupdict()\n",
    "            if mapper['cls'] == \"Gene\":\n",
    "              properties = {\n",
    "                \"id\": genes[group_dict[\"symbol\"]][\"id\"],\n",
    "                \"label\": genes[group_dict[\"symbol\"]][\"label\"]\n",
    "              }\n",
    "              if group_dict[\"symbol\"] in gene_scores:\n",
    "                print(group_dict[\"symbol\"])\n",
    "                for k,v in gene_scores.loc[group_dict[\"symbol\"]]:\n",
    "                  properties[k.replace(\".\", \"_\" )] = v\n",
    "              gene_symbol, gene_id = get_info(group_dict[\"symbol\"])\n",
    "              if gene_symbol and gene_id:\n",
    "                properties[\"label\"] = gene_symbol\n",
    "                properties[\"id\"] = gene_id\n",
    "                properties[\"uri\"] = \"https://www.ncbi.nlm.nih.gov/gene/%s\"%gene_id\n",
    "                \n",
    "                \n",
    "              else:\n",
    "                properties[\"uri\"] = \"https://uswest.ensembl.org/Homo_sapiens/Gene/Summary?g=%s\"%group_dict[\"symbol\"]\n",
    "              return { \"type\": mapper['cls'], \"properties\": {**properties}}\n",
    "            elif mapper['cls'] == \"Drug\":\n",
    "              gd = drugs[group_dict[\"id\"]]\n",
    "              if not str(drugs[group_dict[\"id\"]][\"id\"]).startswith(\"LSM\"):\n",
    "                gd[\"id\"] = str(drugs[group_dict[\"id\"]][\"id\"])\n",
    "                gd[\"uri\"] = \"https://pubchem.ncbi.nlm.nih.gov/compound/%s\"%str(group_dict[\"id\"])\n",
    "              return {\"type\": mapper['cls'], \"properties\": gd}\n",
    "            elif mapper['cls'] == \"BirthDefect\":\n",
    "              birth_defect_id = group_dict[\"id\"]\n",
    "              if birth_defect_id in birth_defects:\n",
    "                birth_defect = birth_defects[birth_defect_id]\n",
    "                properties = {\n",
    "                  \"id\": birth_defect_id,\n",
    "                  \"label\": birth_defect,\n",
    "                  \"uri\": \"https://purl.obolibrary.org/obo/=%s\"%birth_defect\n",
    "                }\n",
    "                return { \"type\": mapper['cls'], \"properties\": {**properties}}\n",
    "              elif birth_defect_id in orpha:\n",
    "                properties = orpha[birth_defect_id]\n",
    "                return { \"type\": mapper['cls'], \"properties\": {**properties}}\n",
    "              else:\n",
    "                res = requests.get(\"https://www.ebi.ac.uk/ols/api/select?q=%s\"%birth_defect_id)\n",
    "                if not res.ok:\n",
    "                  print(\"ERROR %s\"%birth_defect_id)\n",
    "                  print(res.text)\n",
    "                else:\n",
    "                  print(birth_defect_id)\n",
    "                  birth_defect = res.json()[\"response\"][\"docs\"][0][\"label\"]\n",
    "                  birth_defect_id = res.json()[\"response\"][\"docs\"][0][\"obo_id\"]\n",
    "                  properties = {\n",
    "                    \"id\": birth_defect_id,\n",
    "                    \"label\": birth_defect,\n",
    "                    \"uri\": res.json()[\"response\"][\"docs\"][0][\"iri\"]\n",
    "                  }\n",
    "                  orpha[birth_defect_id.replace(\":\", \"_\")] = properties\n",
    "                  return { \"type\": mapper['cls'], \"properties\": {**properties}}\n",
    "            else:\n",
    "              print(group_dict)\n",
    "\n",
    "              \n",
    "              \n",
    "            # if \"id\" in group_dict:\n",
    "            #     if group_dict[\"id\"] in birth_defects:\n",
    "            #         group_dict[\"label\"] = birth_defects[group_dict[\"id\"]]\n",
    "            #         return {\"type\": mapper['cls'], \"properties\": group_dict}\n",
    "            #     elif group_dict[\"id\"] in drugs:\n",
    "            #         gd = drugs[group_dict[\"id\"]]\n",
    "            #         if mapper['cls'] == 'Drug':\n",
    "            #           gd[\"id\"] = \"CID:%s\"%str(drugs[group_dict[\"id\"]][\"id\"])\n",
    "            #         return {\"type\": mapper['cls'], \"properties\": gd}\n",
    "            #     elif group_dict[\"id\"] in relationships:\n",
    "            #         group_dict[\"label\"] = relationships[group_dict[\"id\"]]\n",
    "            #         return {\"type\": mapper['cls'], \"properties\": group_dict}\n",
    "            # elif \"symbol\" in group_dict:\n",
    "            #     gene_symbol, gene_id = get_info(group_dict[\"symbol\"])\n",
    "            #     if gene_symbol and gene_id:\n",
    "            #       properties = genes[group_dict[\"symbol\"]]\n",
    "            #       properties[\"label\"] = gene_symbol\n",
    "            #       properties[\"id\"] = gene_id\n",
    "            #       properties[\"uri\"] = \"https://www.ncbi.nlm.nih.gov/gene/%s\"%gene_id\n",
    "            #       return { \"type\": mapper['cls'], \"properties\": {**group_dict, **properties}}\n",
    "            #     else:\n",
    "            #       return { \"type\": mapper['cls'], \"properties\": {**group_dict, \n",
    "            #         **genes[group_dict[\"symbol\"]],\n",
    "            #         \"uri\": \"https://uswest.ensembl.org/Homo_sapiens/Gene/Summary?g=%s\"%group_dict[\"symbol\"]\n",
    "            #       }}\n",
    "            # return {\"type\": mapper['cls'], \"properties\": group_dict}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'dict' object is not callable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/Users/maayanlab/Documents/codebase/KnowledgExchange/reprotox-kg/notebooks/Serialization.ipynb Cell 11'\u001b[0m in \u001b[0;36m<cell line: 6>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/maayanlab/Documents/codebase/KnowledgExchange/reprotox-kg/notebooks/Serialization.ipynb#ch0000009?line=7'>8</a>\u001b[0m rdfgraph\u001b[39m.\u001b[39mparse(assoc, \u001b[39mformat\u001b[39m\u001b[39m=\u001b[39m\u001b[39m'\u001b[39m\u001b[39mttl\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/maayanlab/Documents/codebase/KnowledgExchange/reprotox-kg/notebooks/Serialization.ipynb#ch0000009?line=8'>9</a>\u001b[0m \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m rdfgraph\u001b[39m.\u001b[39mall_nodes():\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/maayanlab/Documents/codebase/KnowledgExchange/reprotox-kg/notebooks/Serialization.ipynb#ch0000009?line=9'>10</a>\u001b[0m     iri \u001b[39m=\u001b[39m iri_to_node(i)\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/maayanlab/Documents/codebase/KnowledgExchange/reprotox-kg/notebooks/Serialization.ipynb#ch0000009?line=10'>11</a>\u001b[0m     \u001b[39mif\u001b[39;00m iri \u001b[39mand\u001b[39;00m \u001b[39mstr\u001b[39m(iri[\u001b[39m\"\u001b[39m\u001b[39mproperties\u001b[39m\u001b[39m\"\u001b[39m][\u001b[39m\"\u001b[39m\u001b[39mid\u001b[39m\u001b[39m\"\u001b[39m]) \u001b[39mnot\u001b[39;00m \u001b[39min\u001b[39;00m nodes:\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/maayanlab/Documents/codebase/KnowledgExchange/reprotox-kg/notebooks/Serialization.ipynb#ch0000009?line=11'>12</a>\u001b[0m         \u001b[39mif\u001b[39;00m iri[\u001b[39m\"\u001b[39m\u001b[39mtype\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m==\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mDrug\u001b[39m\u001b[39m\"\u001b[39m:\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/maayanlab/Documents/codebase/KnowledgExchange/reprotox-kg/notebooks/Serialization.ipynb#ch0000009?line=12'>13</a>\u001b[0m             \u001b[39m# iri[\"properties\"][\"uri\"] = \"https://pubchem.ncbi.nlm.nih.gov/compound/%s\"%str(iri[\"properties\"][\"id\"].replace(\"CID:\", \"\"))\u001b[39;00m\n",
      "\u001b[1;32m/Users/maayanlab/Documents/codebase/KnowledgExchange/reprotox-kg/notebooks/Serialization.ipynb Cell 10'\u001b[0m in \u001b[0;36miri_to_node\u001b[0;34m(iri)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/maayanlab/Documents/codebase/KnowledgExchange/reprotox-kg/notebooks/Serialization.ipynb#ch0000008?line=48'>49</a>\u001b[0m   \u001b[39mfor\u001b[39;00m k,v \u001b[39min\u001b[39;00m gene_scores\u001b[39m.\u001b[39mloc[group_dict[\u001b[39m\"\u001b[39m\u001b[39msymbol\u001b[39m\u001b[39m\"\u001b[39m]]:\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/maayanlab/Documents/codebase/KnowledgExchange/reprotox-kg/notebooks/Serialization.ipynb#ch0000008?line=49'>50</a>\u001b[0m     properties[k\u001b[39m.\u001b[39mreplace(\u001b[39m\"\u001b[39m\u001b[39m.\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39m_\u001b[39m\u001b[39m\"\u001b[39m )] \u001b[39m=\u001b[39m v\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/maayanlab/Documents/codebase/KnowledgExchange/reprotox-kg/notebooks/Serialization.ipynb#ch0000008?line=50'>51</a>\u001b[0m gene_symbol, gene_id \u001b[39m=\u001b[39m get_info(group_dict[\u001b[39m\"\u001b[39;49m\u001b[39msymbol\u001b[39;49m\u001b[39m\"\u001b[39;49m])\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/maayanlab/Documents/codebase/KnowledgExchange/reprotox-kg/notebooks/Serialization.ipynb#ch0000008?line=51'>52</a>\u001b[0m \u001b[39mif\u001b[39;00m gene_symbol \u001b[39mand\u001b[39;00m gene_id:\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/maayanlab/Documents/codebase/KnowledgExchange/reprotox-kg/notebooks/Serialization.ipynb#ch0000008?line=52'>53</a>\u001b[0m   properties[\u001b[39m\"\u001b[39m\u001b[39mlabel\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m gene_symbol\n",
      "\u001b[1;32m/Users/maayanlab/Documents/codebase/KnowledgExchange/reprotox-kg/notebooks/Serialization.ipynb Cell 2'\u001b[0m in \u001b[0;36mget_info\u001b[0;34m(gene)\u001b[0m\n\u001b[1;32m     <a href='vscode-notebook-cell:/Users/maayanlab/Documents/codebase/KnowledgExchange/reprotox-kg/notebooks/Serialization.ipynb#ch0000001?line=79'>80</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mget_info\u001b[39m(gene):\n\u001b[0;32m---> <a href='vscode-notebook-cell:/Users/maayanlab/Documents/codebase/KnowledgExchange/reprotox-kg/notebooks/Serialization.ipynb#ch0000001?line=80'>81</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m(symbol_mapper(gene), \u001b[39mstr\u001b[39m(id_mapper(gene)))\n",
      "\u001b[0;31mTypeError\u001b[0m: 'dict' object is not callable"
     ]
    }
   ],
   "source": [
    "edges = []\n",
    "nodes = {}\n",
    "hpo_ids = set()\n",
    "no_freq = set()\n",
    "new_ids = {}\n",
    "for assoc, resource in associations.items():\n",
    "    rdfgraph = rdflib.Graph()\n",
    "    rdfgraph.parse(assoc, format='ttl')\n",
    "    for i in rdfgraph.all_nodes():\n",
    "        iri = iri_to_node(i)\n",
    "        if iri and str(iri[\"properties\"][\"id\"]) not in nodes:\n",
    "            if iri[\"type\"] == \"Drug\":\n",
    "                # iri[\"properties\"][\"uri\"] = \"https://pubchem.ncbi.nlm.nih.gov/compound/%s\"%str(iri[\"properties\"][\"id\"].replace(\"CID:\", \"\"))\n",
    "                uid = str(iri[\"properties\"][\"id\"])\n",
    "                label = iri[\"properties\"][\"label\"]\n",
    "                if label in label_id_mapper:\n",
    "                    if uid not in new_ids:\n",
    "                        new_ids[uid] = label_id_mapper[label]\n",
    "                    uid = label_id_mapper[label]\n",
    "                else:\n",
    "                    label_id_mapper[label] = uid\n",
    "                    new_ids[uid] = uid\n",
    "                iri[\"properties\"][\"id\"] = uid\n",
    "                nodes[uid] = {\n",
    "                    **iri,\n",
    "                }\n",
    "            elif iri[\"type\"] == \"Gene\":\n",
    "                uid = str(iri[\"properties\"][\"id\"])\n",
    "                label = iri[\"properties\"][\"label\"]\n",
    "                if label in label_id_mapper:\n",
    "                    if uid not in new_ids:\n",
    "                        new_ids[uid] = label_id_mapper[label]\n",
    "                    uid = label_id_mapper[label]\n",
    "                else:\n",
    "                    label_id_mapper[label] = uid\n",
    "                    new_ids[node_id] = uid\n",
    "                iri[\"properties\"][\"id\"] = uid\n",
    "                nodes[uid] = {\n",
    "                    **iri,\n",
    "                    \n",
    "                }\n",
    "            elif iri[\"type\"] == \"BirthDefect\" and str(iri[\"properties\"][\"id\"].startswith(\"HP:\")):\n",
    "                # iri[\"properties\"][\"uri\"] = \"https://purl.obolibrary.org/obo/=%s\"%str(iri[\"properties\"][\"id\"])\n",
    "                node_id = str(iri[\"properties\"][\"id\"])\n",
    "                label = iri[\"properties\"][\"label\"]\n",
    "                if label in label_id_mapper:\n",
    "                    if node_id not in new_ids:\n",
    "                        new_ids[node_id] = label_id_mapper[label]\n",
    "                    node_id = label_id_mapper[label]\n",
    "                else:\n",
    "                    label_id_mapper[label] = node_id\n",
    "                    new_ids[node_id] = node_id\n",
    "                iri[\"properties\"][\"id\"] = node_id\n",
    "                nodes[node_id] = {\n",
    "                    **iri,\n",
    "                }\n",
    "                if node_id in HPO_Freq.index:\n",
    "                    nodes[node_id][\"KF_freq\"] = int(HPO_Freq.at[str(iri[\"properties\"][\"id\"]), 'KF Frequency'])\n",
    "                    hpo_ids.add(node_id)\n",
    "                else:\n",
    "                    no_freq.add(node_id)\n",
    "            else:\n",
    "                uid = str(iri[\"properties\"][\"id\"])\n",
    "                label = iri[\"properties\"][\"label\"]\n",
    "                if label in label_id_mapper:\n",
    "                    if uid not in new_ids:\n",
    "                        new_ids[uid] = label_id_mapper[label]\n",
    "                    uid = label_id_mapper[label]\n",
    "                else:\n",
    "                    label_id_mapper[label] = uid\n",
    "                    new_ids[uid] = uid\n",
    "                iri[\"properties\"][\"id\"] = uid\n",
    "                nodes[str(iri[\"properties\"][\"id\"])] = {\n",
    "                    **iri,\n",
    "                    \"uri\": str(i)\n",
    "                }\n",
    "    for subj, pred, obj in rdfgraph.triples((None, None, None)):\n",
    "        subj_iri = iri_to_node(subj)\n",
    "        obj_iri = iri_to_node(obj)\n",
    "        if subj_iri and nodes[str(subj_iri[\"properties\"][\"id\"])] and obj_iri and nodes[str(obj_iri[\"properties\"][\"id\"])]:\n",
    "            pred = str(pred)\n",
    "            n = iri_to_node(pred)\n",
    "            if n and n[\"id\"]:\n",
    "                pred = relationships[n[\"id\"]]\n",
    "            \n",
    "            edges.append({\n",
    "                \"type\": \"Relation\",\n",
    "                \"source\": new_ids[str(subj_iri[\"properties\"][\"id\"])],\n",
    "                \"relation\": \"%s\"%resource,#(resource, relationships[pred.split(\"/\")[-1]]),\n",
    "                \"target\": new_ids[str(obj_iri[\"properties\"][\"id\"])],\n",
    "                \"properties\": {\n",
    "                    \"id\": \"%s_%s_%s\"%(subj_iri[\"properties\"][\"label\"], relationships[pred.split(\"/\")[-1]], obj_iri[\"properties\"][\"label\"]),\n",
    "                    \"source_label\": subj_iri[\"properties\"][\"label\"],\n",
    "                    \"target_label\": obj_iri[\"properties\"][\"label\"],\n",
    "                    \"directed\": True,\n",
    "                    \"resource\": resource\n",
    "                }\n",
    "            })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "serialization_v1 = {\n",
    "    \"version\": \"1\",\n",
    "    \"nodes\": nodes,\n",
    "    \"edges\": edges\n",
    "}\n",
    "with open(\"results/reprotox_serialization.v1.json\", \"w\") as o:\n",
    "    o.write(json.dumps(serialization_v1, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def isNumber(value):\n",
    "    try:\n",
    "        v = int(value)\n",
    "        return {\"@type\": \"int\", \"@value\": v}\n",
    "    except:\n",
    "        try:\n",
    "            v = float(value)\n",
    "            return {\"@type\": \"number\", \"@value\": v}\n",
    "        except:\n",
    "            return False\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isinstance('', str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "def typer(value):\n",
    "    numb = isNumber(str(value))\n",
    "    if numb:\n",
    "        return numb\n",
    "    elif isinstance(value, str):\n",
    "        return {\n",
    "            \"@type\": \"string\",\n",
    "            \"@value\": value\n",
    "        }\n",
    "    elif isinstance(value, list):\n",
    "        type_list = []\n",
    "        for i in value:\n",
    "            type_list.append(typer(i))\n",
    "        \n",
    "        return {\n",
    "            \"@type\": \"array\",\n",
    "            \"@value\": type_list\n",
    "        }\n",
    "    elif isinstance(value, dict):\n",
    "        type_dict = {}\n",
    "        for k,v in value.items():\n",
    "            if k == \"id\" or k == \"target\" or k == \"source\":\n",
    "                type_dict[k] = {\n",
    "                    \"@type\": \"string\",\n",
    "                    \"@value\": v\n",
    "                }\n",
    "            else: type_dict[k] = typer(v)\n",
    "        return {\n",
    "            \"@type\": \"object\",\n",
    "            \"@value\": type_dict\n",
    "        }\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_v2 = {}\n",
    "for k,v in nodes.items():\n",
    "    nodes_v2[k] = typer(v)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges_v2 = []\n",
    "for i in edges:\n",
    "    edges_v2.append(\n",
    "        typer(i)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [],
   "source": [
    "serialization_v2 = {\n",
    "    \"version\": \"2\",\n",
    "    \"nodes\": nodes_v2,\n",
    "    \"edges\": edges_v2\n",
    "}\n",
    "with open(\"results/reprotox_serialization.v2.json\", \"w\") as o:\n",
    "    o.write(json.dumps(serialization_v2, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_id(value):\n",
    "    for prefix, val in rdfgraph.namespaces():\n",
    "        v = value.replace(val, \"\")\n",
    "        if not v == value:\n",
    "            return prefix, v\n",
    "    else:\n",
    "        return \"Literal\", value\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SigCom LINCS\n",
    "### drug2gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"data/SigComLINCS_drug_2_gene_25.v1.json\") as o:\n",
    "    sigcom_lincs = json.loads(o.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4420"
      ]
     },
     "execution_count": 228,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_names = set()\n",
    "for k,v in sigcom_lincs[\"nodes\"].items():\n",
    "    t = v[\"type\"]\n",
    "    if t == \"Gene\":\n",
    "        gene_names.add(v[\"properties\"][\"label\"])\n",
    "len(gene_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 229,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4411"
      ]
     },
     "execution_count": 229,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "payload = {\n",
    "    \"filter\": {\n",
    "        \"where\": {\n",
    "            \"meta.symbol\": {\"inq\": list(gene_names)}\n",
    "        }\n",
    "    }\n",
    "}\n",
    "res = requests.post(\"https://maayanlab.cloud/sigcom-lincs/metadata-api/entities/find\", json=payload)\n",
    "len(res.json())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_mapping_name = {\n",
    "    'LINC00341': 'SYNE3',\n",
    "    'ATP5MPL': 'ATP5MJ'\n",
    "}\n",
    "gene_mapping = {\n",
    "    'SYNE3': '161176',\n",
    "    'ATP5MJ': '9556'\n",
    "}\n",
    "for i in res.json():\n",
    "    gene_id = i[\"meta\"][\"geneid\"]\n",
    "    gene_symbol = i[\"meta\"][\"symbol\"]\n",
    "    gene_mapping[gene_symbol] = gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 231,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in gene_names - gene_mapping.keys():\n",
    "    gene_symbol, gene_id = get_info(i)\n",
    "    gene_mapping[gene_symbol] = str(gene_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 232,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FAM30A\n"
     ]
    }
   ],
   "source": [
    "nodes = {}\n",
    "node_map = {}\n",
    "ensembl_map = {}\n",
    "id_mapper = {}\n",
    "for k,v in sigcom_lincs[\"nodes\"].items():\n",
    "    t = v[\"type\"]\n",
    "    v[\"properties\"][\"id\"] = str(v[\"properties\"][\"id\"].replace(\"CID\", \"CID:\"))\n",
    "    new_v = {}\n",
    "    for i,j in v[\"properties\"].items():\n",
    "        new_v[i.replace(\".\", \"_\")] = j\n",
    "    label = new_v[\"label\"]\n",
    "    if t == \"Gene\":\n",
    "        old_id = v[\"properties\"][\"id\"]\n",
    "        gene_label = gene_mapping_name.get(label, label)\n",
    "        gene_id = gene_mapping.get(gene_label)\n",
    "        if gene_label and gene_id:\n",
    "            uid = gene_id\n",
    "            if label in label_id_mapper:\n",
    "                uid = label_id_mapper[label]\n",
    "            else:\n",
    "                label_id_mapper[label] = uid\n",
    "            id_mapper[gene_id] = uid\n",
    "            ensembl_map[old_id] = uid\n",
    "            new_v[\"id\"] = uid\n",
    "            new_v[\"label\"] = gene_label\n",
    "            new_v[\"uri\"] = \"https://www.ncbi.nlm.nih.gov/gene/%s\"%new_v[\"id\"]\n",
    "            label = gene_label\n",
    "            if label in gene_scores.index:\n",
    "                for key,val in gene_scores.loc[label].items():\n",
    "                    new_v[key.replace(\".\", \"_\")] = val\n",
    "        else:\n",
    "            print(label)\n",
    "            continue\n",
    "    elif t == \"Drug\":\n",
    "        uid = new_v[\"id\"]\n",
    "        if label in label_id_mapper:\n",
    "            uid = label_id_mapper[label]\n",
    "        else:\n",
    "            label_id_mapper[label] = uid\n",
    "        id_mapper[new_v[\"id\"]] = uid\n",
    "        new_v[\"id\"] = uid\n",
    "        new_v[\"uri\"] = \"https://pubchem.ncbi.nlm.nih.gov/compound/%s\"%v[\"properties\"][\"id\"].replace(\"CID:\", \"\")\n",
    "        if label in drug_scores.index:\n",
    "            new_v[\"placenta_score\"] = drug_scores.at[label, \"Similarity_Score\"]\n",
    "            new_v[\"placenta_rank\"] = drug_scores.at[label, \"Rank\"]\n",
    "    nodes[new_v[\"id\"]] = {\n",
    "        \"type\": t,\n",
    "        \"properties\": new_v\n",
    "    }\n",
    "    node_map[k] = str(v[\"properties\"][\"id\"])\n",
    "    node_map[str(new_v[\"id\"])] = str(v[\"properties\"][\"id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 233,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges = []\n",
    "not_in = set()\n",
    "in_set = set()\n",
    "for i in sigcom_lincs[\"edges\"]:\n",
    "    source = id_mapper[str(i[\"source\"].replace(\"CID\", \"CID:\"))]\n",
    "    if str(i[\"target\"]) in ensembl_map:\n",
    "        target = id_mapper[ensembl_map[str(i[\"target\"])]]\n",
    "        # if source not in node_map: not_in.add(source)\n",
    "        # else: in_set.add(source)\n",
    "        # if target not in node_map: not_in.add(target)\n",
    "        # else: in_set.add(target)\n",
    "        i[\"source\"] = source\n",
    "        i[\"target\"] = target\n",
    "        source_label = nodes[source][\"properties\"][\"label\"]\n",
    "        target_label = nodes[target][\"properties\"][\"label\"]\n",
    "        i[\"properties\"] = {\n",
    "            \"id\": \"%s %s %s\"%(source_label, i[\"relation\"], target_label),\n",
    "            \"source_label\": source_label,\n",
    "            \"target_label\": target_label,\n",
    "            \"resource\": \"SigCom LINCS\",\n",
    "            \"mean_CD_coefficient\": i.pop(\"weight\"),\n",
    "            \"directed\": True,\n",
    "        }\n",
    "        i[\"relation\"] = \"SigCom LINCS Drug-to-Gene (%s)\"%i[\"relation\"]\n",
    "        edges.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 234,
   "metadata": {},
   "outputs": [],
   "source": [
    "serialization_v1 = {\n",
    "    \"version\": \"1\",\n",
    "    \"nodes\": nodes,\n",
    "    \"edges\": edges\n",
    "}\n",
    "with open(\"results/sigcom_lincs_serialization.v1.json\", \"w\") as o:\n",
    "    o.write(json.dumps(serialization_v1, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_v2 = {}\n",
    "for k,v in nodes.items():\n",
    "    nodes_v2[k] = typer(v)\n",
    "\n",
    "edges_v2 = []\n",
    "for i in edges:\n",
    "    edges_v2.append(\n",
    "        typer(i)\n",
    "    )\n",
    "    \n",
    "serialization_v2 = {\n",
    "    \"version\": \"2\",\n",
    "    \"nodes\": nodes_v2,\n",
    "    \"edges\": edges_v2\n",
    "}\n",
    "with open(\"results/sigcom_lincs_serialization.v2.json\", \"w\") as o:\n",
    "    o.write(json.dumps(serialization_v2, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Gene 2 Drug (Should I do this?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"data/full_weights_SigComLINCS_gene_2_drug07262022.v1.json\") as o:\n",
    "    sigcom_lincs = json.loads(o.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes = {}\n",
    "node_map = {}\n",
    "for k,v in sigcom_lincs[\"nodes\"].items():\n",
    "    t = v[\"type\"]\n",
    "    v[\"properties\"][\"id\"] = str(v[\"properties\"][\"id\"].replace(\"CID\", \"CID:\"))\n",
    "    new_v = {}\n",
    "    for i,j in v[\"properties\"].items():\n",
    "        new_v[i.replace(\".\", \"_\")] = j\n",
    "    label = new_v[\"label\"]\n",
    "    if t == \"Gene\":\n",
    "        new_v[\"id\"] = v[\"properties\"][\"id\"]\n",
    "        new_v[\"label\"] = v[\"properties\"][\"label\"]\n",
    "        new_v[\"uri\"] = \"https://www.ncbi.nlm.nih.gov/gene/%s\"%new_v[\"id\"]\n",
    "        if label in gene_scores.index:\n",
    "            for key,val in gene_scores.loc[label].items():\n",
    "                new_v[key.replace(\".\", \"_\")] = val\n",
    "    elif t == \"Drug\":\n",
    "        new_v[\"uri\"] = \"https://pubchem.ncbi.nlm.nih.gov/compound/%s\"%v[\"properties\"][\"id\"].replace(\"CID:\", \"\")\n",
    "        if label in drug_scores.index:\n",
    "            new_v[\"placenta_score\"] = drug_scores.at[label, \"Similarity_Score\"]\n",
    "            new_v[\"placenta_rank\"] = drug_scores.at[label, \"Rank\"]\n",
    "    nodes[new_v[\"id\"]] = {\n",
    "        \"type\": t,\n",
    "        \"properties\": new_v\n",
    "    }\n",
    "    node_map[k] = str(v[\"properties\"][\"id\"])\n",
    "    node_map[str(new_v[\"id\"])] = str(v[\"properties\"][\"id\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges = []\n",
    "not_in = set()\n",
    "in_set = set()\n",
    "for i in sigcom_lincs[\"edges\"]:\n",
    "    source = str(i[\"source\"].replace(\"CID\", \"CID:\"))\n",
    "    target = str(i[\"target\"])\n",
    "    # if source not in node_map: not_in.add(source)\n",
    "    # else: in_set.add(source)\n",
    "    # if target not in node_map: not_in.add(target)\n",
    "    # else: in_set.add(target)\n",
    "    i[\"source\"] = source\n",
    "    i[\"target\"] = target\n",
    "    source_label = nodes[source][\"properties\"][\"label\"]\n",
    "    target_label = nodes[target][\"properties\"][\"label\"]\n",
    "    i[\"properties\"] = {\n",
    "        \"id\": \"%s %s %s\"%(source_label, i[\"relation\"], target_label),\n",
    "        \"source_label\": source_label,\n",
    "        \"target_label\": target_label,\n",
    "        \"resource\": \"SigCom LINCS\",\n",
    "        \"mean_CD_coefficient\": i.pop(\"weight\"),\n",
    "        \"directed\": True,\n",
    "    }\n",
    "    i[\"relation\"] = \"SigCom LINCS (%s)\"%i[\"relation\"]\n",
    "    edges.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "payload = {\n",
    "    \"filter\": {\n",
    "        \"where\": {\n",
    "            \"meta.symbol\": {\n",
    "                \"inq\": list(not_in)\n",
    "            }\n",
    "        }\n",
    "    }\n",
    "}\n",
    "\n",
    "res = requests.post(\"https://maayanlab.cloud/sigcom-lincs/metadata-api/entities/find\", json=payload)\n",
    "res.ok"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(not_in), len(res.json())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "not_in - set([i[\"meta\"][\"symbol\"] for i in res.json()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n = set()\n",
    "for i in res.json():\n",
    "    if \"ensemblid\" not in i[\"meta\"]:\n",
    "        gene_id = str(i[\"meta\"][\"geneid\"])\n",
    "        label = i[\"meta\"][\"symbol\"]\n",
    "        nodes[gene_id] = {\n",
    "            \"type\": \"Gene\",\n",
    "            \"properties\": {\n",
    "                \"id\": gene_id,\n",
    "                \"label\": label\n",
    "            }\n",
    "        }\n",
    "    else:\n",
    "        gene_id = str(i[\"meta\"][\"ensemblid\"])\n",
    "        label = i[\"meta\"][\"symbol\"]\n",
    "        nodes[gene_id] = {\n",
    "            \"type\": \"Gene\",\n",
    "            \"properties\": {\n",
    "                \"id\": gene_id,\n",
    "                \"label\": label\n",
    "            }\n",
    "        }\n",
    "    node_map[label] = gene_id\n",
    "    node_map[gene_id] = gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sigcom_lincs[\"edges\"][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges = []\n",
    "not_in = {}\n",
    "in_set = set()\n",
    "for i in sigcom_lincs[\"edges\"]:\n",
    "    source = str(i[\"source\"])\n",
    "    target = str(i[\"target\"])\n",
    "    \n",
    "    if source in nodes and target in nodes:\n",
    "        i[\"source\"] = node_map.get(source, source)\n",
    "        i[\"target\"] = node_map.get(target, target)\n",
    "        cd = i[\"properties\"].pop(\"Mean CD-coefficient\", None)\n",
    "        if cd:\n",
    "            i[\"properties\"][\"mean_CD_coefficient\"] = cd\n",
    "            source_label = nodes[i[\"source\"]][\"properties\"][\"label\"]\n",
    "            target_label = nodes[i[\"target\"]][\"properties\"][\"label\"]\n",
    "            i[\"properties\"][\"source_label\"] = source_label\n",
    "            i[\"properties\"][\"target_label\"] = target_label\n",
    "            i[\"properties\"][\"id\"] = \"%s_%s_%s\"%(source_label, i[\"relation\"], target_label)\n",
    "            i[\"relation\"] = \"SigCom LINCS (%s)\"%i[\"relation\"]\n",
    "        edges.append(i)\n",
    "    elif not source in nodes:\n",
    "        if source not in not_in:\n",
    "            not_in[source] = 0\n",
    "        not_in[source]+=1\n",
    "    elif not target in nodes:\n",
    "        if target not in not_in:\n",
    "            not_in[target] = 0\n",
    "        not_in[target]+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(edges), len(sigcom_lincs[\"edges\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "serialization_v1 = {\n",
    "    \"version\": \"1\",\n",
    "    \"nodes\": nodes,\n",
    "    \"edges\": edges\n",
    "}\n",
    "with open(\"results/sigcom_lincs_serialization.v1.json\", \"w\") as o:\n",
    "    o.write(json.dumps(serialization_v1, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_v2 = {}\n",
    "for k,v in nodes.items():\n",
    "    nodes_v2[k] = typer(v)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges_v2 = []\n",
    "for i in edges:\n",
    "    edges_v2.append(\n",
    "        typer(i)\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_v2 = {}\n",
    "for k,v in nodes.items():\n",
    "    nodes_v2[k] = typer(v)\n",
    "\n",
    "edges_v2 = []\n",
    "for i in edges:\n",
    "    edges_v2.append(\n",
    "        typer(i)\n",
    "    )\n",
    "    \n",
    "serialization_v2 = {\n",
    "    \"version\": \"2\",\n",
    "    \"nodes\": nodes_v2,\n",
    "    \"edges\": edges_v2\n",
    "}\n",
    "with open(\"results/sigcom_lincs_serialization.v2.json\", \"w\") as o:\n",
    "    o.write(json.dumps(serialization_v2, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drug2Drug"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"data/full_SigComLINCS_lm_drug_cosine_sim.v1.json\") as o:\n",
    "    drug2drug = json.loads(o.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CID:4549\n"
     ]
    }
   ],
   "source": [
    "nodes = {}\n",
    "id_mapper = {}\n",
    "for k,v in drug2drug[\"nodes\"].items():\n",
    "    properties = v[\"properties\"]\n",
    "    properties[\"id\"] = k.replace(\"CID\", \"CID:\")\n",
    "    if k == \"CID4549\":\n",
    "        print(properties[\"id\"])\n",
    "    label = properties[\"label\"]\n",
    "    uid = properties[\"id\"]\n",
    "    if label in label_id_mapper:\n",
    "        id_mapper[uid] = label_id_mapper[label]\n",
    "        uid = label_id_mapper[label]\n",
    "    else:\n",
    "        id_mapper[uid] = uid\n",
    "        label_id_mapper[label] = uid\n",
    "    properties[\"id\"] = uid\n",
    "\n",
    "    id_mapper[properties[\"id\"]] = uid\n",
    "    properties[\"uri\"] = \"https://pubchem.ncbi.nlm.nih.gov/compound/%s\"%v[\"properties\"][\"id\"].replace(\"CID:\", \"\")\n",
    "    if label in drug_scores.index:\n",
    "        properties[\"placenta_score\"] = drug_scores.at[label, \"Similarity_Score\"]\n",
    "        properties[\"placenta_rank\"] = drug_scores.at[label, \"Rank\"]\n",
    "    nodes[properties[\"id\"]] = {\n",
    "        \"type\": v[\"type\"],\n",
    "        \"properties\": properties\n",
    "    }\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "edges = []\n",
    "for i in drug2drug[\"edges\"]:\n",
    "    source = id_mapper[i[\"source\"].replace(\"CID\", \"CID:\")]\n",
    "    source_label = nodes[source][\"properties\"][\"label\"]\n",
    "    target = id_mapper[i[\"target\"].replace(\"CID\", \"CID:\")]\n",
    "    target_label = nodes[target][\"properties\"][\"label\"]\n",
    "    relation = \"LINCS Drugs Cosine Similarity\"\n",
    "    cosine_similarity = i[\"weight\"]\n",
    "\n",
    "    edges.append({\n",
    "        \"source\": source,\n",
    "        \"relation\": relation,\n",
    "        \"target\": target,\n",
    "        \"properties\": {\n",
    "            \"id\": \"%s-%s similarity\"%(source_label, target_label),\n",
    "            \"source_label\": source_label,\n",
    "            \"target_label\": target_label,\n",
    "            \"cosine_similarity\": cosine_similarity\n",
    "        }\n",
    "    })\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'type': 'Drug',\n",
       " 'properties': {'id': 'CID:4549',\n",
       "  'label': 'NPPB',\n",
       "  'uri': 'https://pubchem.ncbi.nlm.nih.gov/compound/4549',\n",
       "  'placenta_score': 0.016815556,\n",
       "  'placenta_rank': 21011.0}}"
      ]
     },
     "execution_count": 145,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug2drug[\"nodes\"]['CID4549']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "serialization_v1 = {\n",
    "    \"version\": \"1\",\n",
    "    \"nodes\": nodes,\n",
    "    \"edges\": edges\n",
    "}\n",
    "with open(\"results/sigcom_lincs_drug_similarity.v1.json\", \"w\") as o:\n",
    "    o.write(json.dumps(serialization_v1, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_v2 = {}\n",
    "for k,v in nodes.items():\n",
    "    nodes_v2[k] = typer(v)\n",
    "\n",
    "edges_v2 = []\n",
    "for i in edges:\n",
    "    edges_v2.append(\n",
    "        typer(i)\n",
    "    )\n",
    "\n",
    "serialization_v2 = {\n",
    "    \"version\": \"2\",\n",
    "    \"nodes\": nodes_v2,\n",
    "    \"edges\": edges_v2\n",
    "}\n",
    "with open(\"results/sigcom_lincs_drug_similarity.v2.json\", \"w\") as o:\n",
    "    o.write(json.dumps(serialization_v2, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Drug Target"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DRUG_SMILES</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>DRUG_ID</th>\n",
       "      <th>DRUG_PUBCHEM_CID</th>\n",
       "      <th>TARGET_NAME</th>\n",
       "      <th>TARGET_CLASS</th>\n",
       "      <th>TARGET_GENE</th>\n",
       "      <th>TARGET_ACCESSION</th>\n",
       "      <th>TARGET_SWISSPROT</th>\n",
       "      <th>TARGET_ORGANISM</th>\n",
       "      <th>TARGET_TDL</th>\n",
       "      <th>ACT_TYPE</th>\n",
       "      <th>ACT_RELATION</th>\n",
       "      <th>ACT_VALUE</th>\n",
       "      <th>ACT_COMMENT</th>\n",
       "      <th>ACT_SOURCE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CCN(CC)CCNC(=O)C1=C(C)NC(\\C=C2/C(=O)NC3=C2C=C(...</td>\n",
       "      <td>sunitinib</td>\n",
       "      <td>2544</td>\n",
       "      <td>5329102</td>\n",
       "      <td>AP2-associated protein kinase 1</td>\n",
       "      <td>Kinase</td>\n",
       "      <td>AAK1</td>\n",
       "      <td>Q2M2I8</td>\n",
       "      <td>AAK1_HUMAN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Tchem</td>\n",
       "      <td>Kd</td>\n",
       "      <td>=</td>\n",
       "      <td>7.96</td>\n",
       "      <td>Binding constant for AAK1 kinase domain</td>\n",
       "      <td>CHEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CN(C)C[C@@H]1CCN2C=C(C3=CC=CC=C23)C2=C(C(=O)NC...</td>\n",
       "      <td>ruboxistaurin</td>\n",
       "      <td>3533</td>\n",
       "      <td>153999</td>\n",
       "      <td>AP2-associated protein kinase 1</td>\n",
       "      <td>Kinase</td>\n",
       "      <td>AAK1</td>\n",
       "      <td>Q2M2I8</td>\n",
       "      <td>AAK1_HUMAN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Tchem</td>\n",
       "      <td>Kd</td>\n",
       "      <td>=</td>\n",
       "      <td>6.05</td>\n",
       "      <td>Binding constant for AAK1 kinase domain</td>\n",
       "      <td>CHEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>N#CC[C@H](C1CCCC1)N1C=C(C=N1)C1=C2C=CNC2=NC=N1</td>\n",
       "      <td>ruxolitinib</td>\n",
       "      <td>4190</td>\n",
       "      <td>25126798</td>\n",
       "      <td>AP2-associated protein kinase 1</td>\n",
       "      <td>Kinase</td>\n",
       "      <td>AAK1</td>\n",
       "      <td>Q2M2I8</td>\n",
       "      <td>AAK1_HUMAN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Tchem</td>\n",
       "      <td>Kd</td>\n",
       "      <td>=</td>\n",
       "      <td>6.92</td>\n",
       "      <td>Binding constant for AAK1 kinase domain</td>\n",
       "      <td>CHEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>COC(=O)C1=CC2=C(C=C1)\\C(=C(\\NC1=CC=C(C=C1)N(C)...</td>\n",
       "      <td>nintedanib</td>\n",
       "      <td>4903</td>\n",
       "      <td>135423438</td>\n",
       "      <td>AP2-associated protein kinase 1</td>\n",
       "      <td>Kinase</td>\n",
       "      <td>AAK1</td>\n",
       "      <td>Q2M2I8</td>\n",
       "      <td>AAK1_HUMAN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Tchem</td>\n",
       "      <td>Kd</td>\n",
       "      <td>=</td>\n",
       "      <td>7.20</td>\n",
       "      <td>Binding constant for AAK1 kinase domain</td>\n",
       "      <td>CHEMBL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CCS(=O)(=O)N1CC(CC#N)(C1)N1C=C(C=N1)C1=NC=NC2=...</td>\n",
       "      <td>baricitinib</td>\n",
       "      <td>5202</td>\n",
       "      <td>44205240</td>\n",
       "      <td>AP2-associated protein kinase 1</td>\n",
       "      <td>Kinase</td>\n",
       "      <td>AAK1</td>\n",
       "      <td>Q2M2I8</td>\n",
       "      <td>AAK1_HUMAN</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Tchem</td>\n",
       "      <td>Kd</td>\n",
       "      <td>=</td>\n",
       "      <td>7.77</td>\n",
       "      <td>Binding affinity determined in a cell-free bio...</td>\n",
       "      <td>IUPHAR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         DRUG_SMILES      DRUG_NAME  DRUG_ID  \\\n",
       "0  CCN(CC)CCNC(=O)C1=C(C)NC(\\C=C2/C(=O)NC3=C2C=C(...      sunitinib     2544   \n",
       "1  CN(C)C[C@@H]1CCN2C=C(C3=CC=CC=C23)C2=C(C(=O)NC...  ruboxistaurin     3533   \n",
       "2     N#CC[C@H](C1CCCC1)N1C=C(C=N1)C1=C2C=CNC2=NC=N1    ruxolitinib     4190   \n",
       "3  COC(=O)C1=CC2=C(C=C1)\\C(=C(\\NC1=CC=C(C=C1)N(C)...     nintedanib     4903   \n",
       "4  CCS(=O)(=O)N1CC(CC#N)(C1)N1C=C(C=N1)C1=NC=NC2=...    baricitinib     5202   \n",
       "\n",
       "   DRUG_PUBCHEM_CID                      TARGET_NAME TARGET_CLASS TARGET_GENE  \\\n",
       "0           5329102  AP2-associated protein kinase 1       Kinase        AAK1   \n",
       "1            153999  AP2-associated protein kinase 1       Kinase        AAK1   \n",
       "2          25126798  AP2-associated protein kinase 1       Kinase        AAK1   \n",
       "3         135423438  AP2-associated protein kinase 1       Kinase        AAK1   \n",
       "4          44205240  AP2-associated protein kinase 1       Kinase        AAK1   \n",
       "\n",
       "  TARGET_ACCESSION TARGET_SWISSPROT TARGET_ORGANISM TARGET_TDL ACT_TYPE  \\\n",
       "0           Q2M2I8       AAK1_HUMAN    Homo sapiens      Tchem       Kd   \n",
       "1           Q2M2I8       AAK1_HUMAN    Homo sapiens      Tchem       Kd   \n",
       "2           Q2M2I8       AAK1_HUMAN    Homo sapiens      Tchem       Kd   \n",
       "3           Q2M2I8       AAK1_HUMAN    Homo sapiens      Tchem       Kd   \n",
       "4           Q2M2I8       AAK1_HUMAN    Homo sapiens      Tchem       Kd   \n",
       "\n",
       "  ACT_RELATION  ACT_VALUE                                        ACT_COMMENT  \\\n",
       "0            =       7.96            Binding constant for AAK1 kinase domain   \n",
       "1            =       6.05            Binding constant for AAK1 kinase domain   \n",
       "2            =       6.92            Binding constant for AAK1 kinase domain   \n",
       "3            =       7.20            Binding constant for AAK1 kinase domain   \n",
       "4            =       7.77  Binding affinity determined in a cell-free bio...   \n",
       "\n",
       "  ACT_SOURCE  \n",
       "0     CHEMBL  \n",
       "1     CHEMBL  \n",
       "2     CHEMBL  \n",
       "3     CHEMBL  \n",
       "4     IUPHAR  "
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_target = pd.read_csv(\"data/idg_target.tsv\", sep=\"\\t\")\n",
    "drug_target.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes = {}\n",
    "edges = []"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MT-CO2\n",
      "MT-CO2\n"
     ]
    }
   ],
   "source": [
    "for drug_id in set(drug_target[\"DRUG_ID\"]):\n",
    "    sub_df = drug_target[drug_target[\"DRUG_ID\"] == drug_id]\n",
    "    for i in sub_df.index:\n",
    "        drug_pubchem_id = \"CID:%s\"%sub_df.at[i, \"DRUG_PUBCHEM_CID\"]\n",
    "        drug_name = sub_df.at[i, \"DRUG_NAME\"]\n",
    "        if drug_name in label_id_mapper:\n",
    "            drug_pubchem_id = label_id_mapper[drug_name]\n",
    "        else:\n",
    "            label_id_mapper[drug_name] = drug_pubchem_id\n",
    "        if drug_pubchem_id not in nodes:\n",
    "            properties = {\n",
    "                \"id\": drug_pubchem_id,\n",
    "                \"label\": drug_name,\n",
    "                \"smiles\": sub_df.at[i, \"DRUG_SMILES\"],\n",
    "                \"IDG_ID\": str(sub_df.at[i, \"DRUG_ID\"]),\n",
    "                \"uri\": \"https://pubchem.ncbi.nlm.nih.gov/compound/%s\"%str(sub_df.at[i, \"DRUG_PUBCHEM_CID\"])\n",
    "            }\n",
    "            if drug_name in drug_scores.index:\n",
    "                properties[\"placenta_score\"] = float(drug_scores.at[drug_name, \"Similarity_Score\"])\n",
    "                properties[\"placenta_rank\"] = int(drug_scores.at[drug_name, \"Rank\"])\n",
    "            nodes[drug_pubchem_id] = {\n",
    "                \"type\": \"Drug\",\n",
    "                \"properties\": properties\n",
    "            }\n",
    "        \n",
    "        gene_label = sub_df.at[i, \"TARGET_GENE\"]\n",
    "        gene_label, gene_id = get_info(gene_label)\n",
    "        if gene_label and gene_id:\n",
    "            if gene_id not in nodes:\n",
    "                properties = {\n",
    "                    \"id\": gene_id,\n",
    "                    \"label\": gene_label,\n",
    "                    \"target_name\": str(sub_df.at[i, \"TARGET_NAME\"]),\n",
    "                    \"target_class\": str(sub_df.at[i, \"TARGET_CLASS\"]),\n",
    "                    \"accession\": str(sub_df.at[i, \"TARGET_ACCESSION\"]),\n",
    "                    \"swissprot\": str(sub_df.at[i, \"TARGET_SWISSPROT\"]),\n",
    "                    \"organism\": str(sub_df.at[i, \"TARGET_ORGANISM\"]),\n",
    "                    \"TDL\": str(sub_df.at[i, \"TARGET_TDL\"])\n",
    "                }\n",
    "                if gene_label in gene_scores.index:\n",
    "                    for k,v in gene_scores.loc[gene_label].items():\n",
    "                        properties[k.replace(\".\", \"_\")] = float(v)\n",
    "                nodes[gene_id] = {\n",
    "                    \"type\": \"Gene\",\n",
    "                    \"properties\": properties\n",
    "                }\n",
    "            edge = {\n",
    "                \"source\": drug_pubchem_id,\n",
    "                \"relation\": \"IDG (Drug Target)\",\n",
    "                \"target\": gene_id,\n",
    "                \"properties\": {\n",
    "                    \"id\": \"%s targets %s\"%(drug_name, gene_label),\n",
    "                    \"source_label\": drug_name,\n",
    "                    \"target_label\": gene_label,\n",
    "                    \"act_type\": str(sub_df.at[i, \"ACT_TYPE\"]),\n",
    "                    \"act_relation\": str(sub_df.at[i, \"ACT_RELATION\"]),\n",
    "                    \"act_value\": str(sub_df.at[i, \"ACT_VALUE\"]),\n",
    "                    \"act_comment\": str(sub_df.at[i, \"ACT_COMMENT\"]),\n",
    "                    \"act_source\": str(sub_df.at[i, \"ACT_SOURCE\"]),\n",
    "                }\n",
    "            }\n",
    "            edges.append(edge)\n",
    "        else:\n",
    "            print(sub_df.at[i, \"TARGET_GENE\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(7326, 16)"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "drug_target.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "serialization_v1 = {\n",
    "    \"version\": \"1\",\n",
    "    \"nodes\": nodes,\n",
    "    \"edges\": edges\n",
    "}\n",
    "with open(\"results/idg_drug_targets.v1.json\", \"w\") as o:\n",
    "    o.write(json.dumps(serialization_v1, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_v2 = {}\n",
    "for k,v in nodes.items():\n",
    "    nodes_v2[k] = typer(v)\n",
    "\n",
    "edges_v2 = []\n",
    "for i in edges:\n",
    "    edges_v2.append(\n",
    "        typer(i)\n",
    "    )\n",
    "\n",
    "serialization_v2 = {\n",
    "    \"version\": \"2\",\n",
    "    \"nodes\": nodes_v2,\n",
    "    \"edges\": edges_v2\n",
    "}\n",
    "with open(\"results/idg_drug_targets.v2.json\", \"w\") as o:\n",
    "    o.write(json.dumps(serialization_v2, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### HPO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "HPO_Freq.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes = {}\n",
    "edges = []\n",
    "for i in HPO_Freq.index:\n",
    "    if i not in nodes:\n",
    "        label = HPO_Freq.at[i, \"HPO description\"]\n",
    "        if type(label) == pd.core.series.Series:\n",
    "            label = label[0]\n",
    "        KF_Freq =  HPO_Freq.at[i, \"KF Frequency\"]\n",
    "        if type(KF_Freq) == pd.core.series.Series:\n",
    "            KF_Freq = int(KF_Freq[0])\n",
    "        else:\n",
    "            KF_Freq = int(KF_Freq)\n",
    "        nodes[i] = {\n",
    "            \"type\": \"BirthDefect\",\n",
    "            \"properties\": {\n",
    "                \"id\": i,\n",
    "                \"label\": label,\n",
    "                \"KF_Freq\": KF_Freq,\n",
    "                \"uri\": \"https://purl.obolibrary.org/obo/%s\"%i.replace(\":\", \"_\")\n",
    "            }\n",
    "        }\n",
    "        time.sleep(0.1)\n",
    "        res = requests.get(\"https://hpo.jax.org/api/hpo/term/%s/genes?max=-1&offset=1\"%i, verify=False)\n",
    "        if not res.ok:\n",
    "            print(i)\n",
    "        else:\n",
    "            for gene_info in res.json()[\"genes\"]:\n",
    "                entrezGeneId = str(gene_info[\"entrezGeneId\"])\n",
    "                entrezGeneSymbol = gene_info[\"entrezGeneSymbol\"]\n",
    "                if entrezGeneId not in nodes:\n",
    "                    properties = {\n",
    "                        \"id\": entrezGeneId,\n",
    "                        \"label\": entrezGeneSymbol,\n",
    "                        \"uri\": \"https://www.ncbi.nlm.nih.gov/gene/%s\"%entrezGeneId\n",
    "                    }\n",
    "                    if entrezGeneSymbol in gene_scores:\n",
    "                        for k,v in gene_scores.loc[entrezGeneSymbol]:\n",
    "                            properties[k.replace(\".\", \"_\" )] = v\n",
    "                    \n",
    "                    nodes[entrezGeneId] = {\n",
    "                        \"type\": \"Gene\",\n",
    "                        \"properties\": properties\n",
    "                    }\n",
    "                edge = {\n",
    "                    \"source\": i,\n",
    "                    \"relation\": \"HPO\",\n",
    "                    \"target\": entrezGeneId,\n",
    "                    \"properties\": {\n",
    "                        \"id\": \"%s (%s-%s)\"%(i, label, entrezGeneSymbol),\n",
    "                        \"source_label\": label,\n",
    "                        \"target_label\": entrezGeneSymbol,\n",
    "                        \"resource\": \"HPO\"\n",
    "                    }\n",
    "                }\n",
    "                edges.append(edge)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "serialization_v1 = {\n",
    "    \"version\": \"1\",\n",
    "    \"nodes\": nodes,\n",
    "    \"edges\": edges\n",
    "}\n",
    "with open(\"results/hpo.v1.json\", \"w\") as o:\n",
    "    o.write(json.dumps(serialization_v1, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "nodes_v2 = {}\n",
    "for k,v in nodes.items():\n",
    "    nodes_v2[k] = typer(v)\n",
    "\n",
    "edges_v2 = []\n",
    "for i in edges:\n",
    "    edges_v2.append(\n",
    "        typer(i)\n",
    "    )\n",
    "\n",
    "serialization_v2 = {\n",
    "    \"version\": \"2\",\n",
    "    \"nodes\": nodes_v2,\n",
    "    \"edges\": edges_v2\n",
    "}\n",
    "with open(\"results/hpo.v2.json\", \"w\") as o:\n",
    "    o.write(json.dumps(serialization_v2, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in glob.glob(\"results/*.v1.json\"):\n",
    "    with open(i) as o:\n",
    "        with open(i.replace(\"results\", \"ingestion\").replace(\"v1\", \"valid\"), 'w') as w:\n",
    "            w.write(o.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NPPB CID:4549 4879\n"
     ]
    }
   ],
   "source": [
    "duplicates = {}\n",
    "for i in glob.glob(\"results/*.v1.json\"):\n",
    "    with open(i) as o:\n",
    "        serialization = json.loads(o.read())\n",
    "        for k,v in serialization[\"nodes\"].items():\n",
    "            uid = v[\"properties\"][\"id\"]\n",
    "            label = v[\"properties\"][\"label\"]\n",
    "            if label in duplicates and not duplicates[label] == uid:\n",
    "                print(label, uid, duplicates[label])\n",
    "            else:\n",
    "                duplicates[label] = uid\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ARCHS4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyarrow.feather as feather\n",
    "from scipy.stats import rankdata, zscore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_correlation = feather.read_feather('data/human_correlation_archs4.f')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AACSP1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.311017</td>\n",
       "      <td>0.074197</td>\n",
       "      <td>0.011767</td>\n",
       "      <td>0.015465</td>\n",
       "      <td>-0.083539</td>\n",
       "      <td>0.024190</td>\n",
       "      <td>-0.023043</td>\n",
       "      <td>0.116815</td>\n",
       "      <td>0.005853</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.020775</td>\n",
       "      <td>-0.065653</td>\n",
       "      <td>-0.034081</td>\n",
       "      <td>-0.002724</td>\n",
       "      <td>-0.022020</td>\n",
       "      <td>0.101489</td>\n",
       "      <td>0.020808</td>\n",
       "      <td>-0.093609</td>\n",
       "      <td>-0.025960</td>\n",
       "      <td>0.003271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0.311017</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.314577</td>\n",
       "      <td>-0.001876</td>\n",
       "      <td>-0.000682</td>\n",
       "      <td>-0.065426</td>\n",
       "      <td>0.020387</td>\n",
       "      <td>-0.047918</td>\n",
       "      <td>0.046401</td>\n",
       "      <td>-0.004944</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.079724</td>\n",
       "      <td>-0.071680</td>\n",
       "      <td>-0.018917</td>\n",
       "      <td>0.047886</td>\n",
       "      <td>0.003180</td>\n",
       "      <td>0.002353</td>\n",
       "      <td>-0.007614</td>\n",
       "      <td>-0.086449</td>\n",
       "      <td>0.018525</td>\n",
       "      <td>0.003854</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.074197</td>\n",
       "      <td>0.314577</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.028479</td>\n",
       "      <td>-0.017056</td>\n",
       "      <td>0.014158</td>\n",
       "      <td>0.014082</td>\n",
       "      <td>-0.021786</td>\n",
       "      <td>-0.059868</td>\n",
       "      <td>-0.022951</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.097068</td>\n",
       "      <td>-0.067448</td>\n",
       "      <td>0.001960</td>\n",
       "      <td>0.018967</td>\n",
       "      <td>0.046186</td>\n",
       "      <td>-0.111606</td>\n",
       "      <td>-0.013873</td>\n",
       "      <td>0.028855</td>\n",
       "      <td>0.021196</td>\n",
       "      <td>-0.037100</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>0.011767</td>\n",
       "      <td>-0.001876</td>\n",
       "      <td>-0.028479</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.007315</td>\n",
       "      <td>0.038877</td>\n",
       "      <td>-0.005643</td>\n",
       "      <td>-0.024810</td>\n",
       "      <td>0.058989</td>\n",
       "      <td>0.031632</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.021674</td>\n",
       "      <td>-0.047888</td>\n",
       "      <td>0.005581</td>\n",
       "      <td>0.009300</td>\n",
       "      <td>-0.008702</td>\n",
       "      <td>0.031462</td>\n",
       "      <td>0.013790</td>\n",
       "      <td>-0.055190</td>\n",
       "      <td>0.001249</td>\n",
       "      <td>0.018487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2MP1</th>\n",
       "      <td>0.015465</td>\n",
       "      <td>-0.000682</td>\n",
       "      <td>-0.017056</td>\n",
       "      <td>0.007315</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>-0.035422</td>\n",
       "      <td>-0.008135</td>\n",
       "      <td>-0.002369</td>\n",
       "      <td>-0.003118</td>\n",
       "      <td>0.011400</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.039716</td>\n",
       "      <td>-0.023186</td>\n",
       "      <td>0.007086</td>\n",
       "      <td>-0.010465</td>\n",
       "      <td>0.026512</td>\n",
       "      <td>0.061912</td>\n",
       "      <td>0.003802</td>\n",
       "      <td>-0.027152</td>\n",
       "      <td>0.010331</td>\n",
       "      <td>0.035132</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  26415 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           A1BG      A1CF       A2M     A2ML1     A2MP1    A4GALT     A4GNT  \\\n",
       "A1BG   1.000000  0.311017  0.074197  0.011767  0.015465 -0.083539  0.024190   \n",
       "A1CF   0.311017  1.000000  0.314577 -0.001876 -0.000682 -0.065426  0.020387   \n",
       "A2M    0.074197  0.314577  1.000000 -0.028479 -0.017056  0.014158  0.014082   \n",
       "A2ML1  0.011767 -0.001876 -0.028479  1.000000  0.007315  0.038877 -0.005643   \n",
       "A2MP1  0.015465 -0.000682 -0.017056  0.007315  1.000000 -0.035422 -0.008135   \n",
       "\n",
       "           AAAS      AACS    AACSP1  ...    ZWILCH     ZWINT      ZXDA  \\\n",
       "A1BG  -0.023043  0.116815  0.005853  ... -0.020775 -0.065653 -0.034081   \n",
       "A1CF  -0.047918  0.046401 -0.004944  ... -0.079724 -0.071680 -0.018917   \n",
       "A2M   -0.021786 -0.059868 -0.022951  ... -0.097068 -0.067448  0.001960   \n",
       "A2ML1 -0.024810  0.058989  0.031632  ... -0.021674 -0.047888  0.005581   \n",
       "A2MP1 -0.002369 -0.003118  0.011400  ... -0.039716 -0.023186  0.007086   \n",
       "\n",
       "           ZXDB      ZXDC    ZYG11A    ZYG11B       ZYX     ZZEF1      ZZZ3  \n",
       "A1BG  -0.002724 -0.022020  0.101489  0.020808 -0.093609 -0.025960  0.003271  \n",
       "A1CF   0.047886  0.003180  0.002353 -0.007614 -0.086449  0.018525  0.003854  \n",
       "A2M    0.018967  0.046186 -0.111606 -0.013873  0.028855  0.021196 -0.037100  \n",
       "A2ML1  0.009300 -0.008702  0.031462  0.013790 -0.055190  0.001249  0.018487  \n",
       "A2MP1 -0.010465  0.026512  0.061912  0.003802 -0.027152  0.010331  0.035132  \n",
       "\n",
       "[5 rows x 26415 columns]"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_correlation.index = human_correlation.columns\n",
    "human_correlation.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_correlation_rank = pd.DataFrame(0, columns=human_correlation.columns, index=human_correlation.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "mat = np.zeros(shape=(len(human_correlation.columns), len(human_correlation.columns)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 186,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|| 26415/26415 [01:24<00:00, 311.73it/s]\n"
     ]
    }
   ],
   "source": [
    "index = 0\n",
    "for col in tqdm(human_correlation.columns):\n",
    "    mat[index] = rankdata(human_correlation[col], method=\"ordinal\")\n",
    "    index += 1\n",
    "\n",
    "human_correlation_rank = pd.DataFrame(mat.T, columns=human_correlation.columns, index=human_correlation.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 187,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A2ML1</th>\n",
       "      <th>A2MP1</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>A4GNT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AACSP1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11A</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>26415.0</td>\n",
       "      <td>26139.0</td>\n",
       "      <td>24341.0</td>\n",
       "      <td>19615.0</td>\n",
       "      <td>17890.0</td>\n",
       "      <td>1504.0</td>\n",
       "      <td>24806.0</td>\n",
       "      <td>7732.0</td>\n",
       "      <td>24667.0</td>\n",
       "      <td>15673.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8400.0</td>\n",
       "      <td>3239.0</td>\n",
       "      <td>3643.0</td>\n",
       "      <td>15276.0</td>\n",
       "      <td>8709.0</td>\n",
       "      <td>23701.0</td>\n",
       "      <td>18611.0</td>\n",
       "      <td>4081.0</td>\n",
       "      <td>3907.0</td>\n",
       "      <td>14736.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>26213.0</td>\n",
       "      <td>26415.0</td>\n",
       "      <td>26290.0</td>\n",
       "      <td>11596.0</td>\n",
       "      <td>12135.0</td>\n",
       "      <td>3123.0</td>\n",
       "      <td>24412.0</td>\n",
       "      <td>2522.0</td>\n",
       "      <td>21136.0</td>\n",
       "      <td>10264.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1906.0</td>\n",
       "      <td>2680.0</td>\n",
       "      <td>8014.0</td>\n",
       "      <td>24413.0</td>\n",
       "      <td>17256.0</td>\n",
       "      <td>10458.0</td>\n",
       "      <td>12104.0</td>\n",
       "      <td>4600.0</td>\n",
       "      <td>21128.0</td>\n",
       "      <td>14911.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>23681.0</td>\n",
       "      <td>26148.0</td>\n",
       "      <td>26415.0</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>4307.0</td>\n",
       "      <td>19063.0</td>\n",
       "      <td>23302.0</td>\n",
       "      <td>8125.0</td>\n",
       "      <td>1845.0</td>\n",
       "      <td>3078.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1186.0</td>\n",
       "      <td>3072.0</td>\n",
       "      <td>16801.0</td>\n",
       "      <td>21126.0</td>\n",
       "      <td>23215.0</td>\n",
       "      <td>757.0</td>\n",
       "      <td>10281.0</td>\n",
       "      <td>20651.0</td>\n",
       "      <td>21551.0</td>\n",
       "      <td>3505.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>14082.0</td>\n",
       "      <td>15952.0</td>\n",
       "      <td>8867.0</td>\n",
       "      <td>26415.0</td>\n",
       "      <td>15361.0</td>\n",
       "      <td>21329.0</td>\n",
       "      <td>9210.0</td>\n",
       "      <td>7215.0</td>\n",
       "      <td>22181.0</td>\n",
       "      <td>22530.0</td>\n",
       "      <td>...</td>\n",
       "      <td>8213.0</td>\n",
       "      <td>5514.0</td>\n",
       "      <td>17860.0</td>\n",
       "      <td>19058.0</td>\n",
       "      <td>13515.0</td>\n",
       "      <td>16049.0</td>\n",
       "      <td>17304.0</td>\n",
       "      <td>7903.0</td>\n",
       "      <td>16350.0</td>\n",
       "      <td>18637.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2MP1</th>\n",
       "      <td>15198.0</td>\n",
       "      <td>16392.0</td>\n",
       "      <td>12952.0</td>\n",
       "      <td>17723.0</td>\n",
       "      <td>26415.0</td>\n",
       "      <td>7577.0</td>\n",
       "      <td>6113.0</td>\n",
       "      <td>15240.0</td>\n",
       "      <td>10237.0</td>\n",
       "      <td>17718.0</td>\n",
       "      <td>...</td>\n",
       "      <td>5174.0</td>\n",
       "      <td>10441.0</td>\n",
       "      <td>18232.0</td>\n",
       "      <td>11824.0</td>\n",
       "      <td>21397.0</td>\n",
       "      <td>20649.0</td>\n",
       "      <td>15113.0</td>\n",
       "      <td>13100.0</td>\n",
       "      <td>19301.0</td>\n",
       "      <td>21531.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  26415 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          A1BG     A1CF      A2M    A2ML1    A2MP1   A4GALT    A4GNT     AAAS  \\\n",
       "A1BG   26415.0  26139.0  24341.0  19615.0  17890.0   1504.0  24806.0   7732.0   \n",
       "A1CF   26213.0  26415.0  26290.0  11596.0  12135.0   3123.0  24412.0   2522.0   \n",
       "A2M    23681.0  26148.0  26415.0   2019.0   4307.0  19063.0  23302.0   8125.0   \n",
       "A2ML1  14082.0  15952.0   8867.0  26415.0  15361.0  21329.0   9210.0   7215.0   \n",
       "A2MP1  15198.0  16392.0  12952.0  17723.0  26415.0   7577.0   6113.0  15240.0   \n",
       "\n",
       "          AACS   AACSP1  ...  ZWILCH    ZWINT     ZXDA     ZXDB     ZXDC  \\\n",
       "A1BG   24667.0  15673.0  ...  8400.0   3239.0   3643.0  15276.0   8709.0   \n",
       "A1CF   21136.0  10264.0  ...  1906.0   2680.0   8014.0  24413.0  17256.0   \n",
       "A2M     1845.0   3078.0  ...  1186.0   3072.0  16801.0  21126.0  23215.0   \n",
       "A2ML1  22181.0  22530.0  ...  8213.0   5514.0  17860.0  19058.0  13515.0   \n",
       "A2MP1  10237.0  17718.0  ...  5174.0  10441.0  18232.0  11824.0  21397.0   \n",
       "\n",
       "        ZYG11A   ZYG11B      ZYX    ZZEF1     ZZZ3  \n",
       "A1BG   23701.0  18611.0   4081.0   3907.0  14736.0  \n",
       "A1CF   10458.0  12104.0   4600.0  21128.0  14911.0  \n",
       "A2M      757.0  10281.0  20651.0  21551.0   3505.0  \n",
       "A2ML1  16049.0  17304.0   7903.0  16350.0  18637.0  \n",
       "A2MP1  20649.0  15113.0  13100.0  19301.0  21531.0  \n",
       "\n",
       "[5 rows x 26415 columns]"
      ]
     },
     "execution_count": 187,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_correlation_rank.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 188,
   "metadata": {},
   "outputs": [],
   "source": [
    "organism='Mammalia/Homo_sapiens'\n",
    "ncbi = fetch_save_read(\n",
    "    'ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/{}.gene_info.gz'.format(organism),\n",
    "    '{}.gene_info.tsv'.format(organism),\n",
    "    sep='\\t',\n",
    ")\n",
    "ncbi = ncbi.set_index(\"GeneID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>-</td>\n",
       "      <td>A1B|ABG|GAB|HYST2477</td>\n",
       "      <td>MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410...</td>\n",
       "      <td>19</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...</td>\n",
       "      <td>20220703</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606</td>\n",
       "      <td>A2M</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MD|CPAMD5|FWP007|S863-7</td>\n",
       "      <td>MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899...</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-2-macroglobulin|C3 and PZP-like alpha-2-...</td>\n",
       "      <td>20220703</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MP</td>\n",
       "      <td>HGNC:HGNC:8|Ensembl:ENSG00000256069|AllianceGe...</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>O</td>\n",
       "      <td>pregnancy-zone protein pseudogene</td>\n",
       "      <td>20220513</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9606</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC1|MNAT|NAT-1|NATI</td>\n",
       "      <td>MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171...</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 1|N-acetyltransf...</td>\n",
       "      <td>20220522</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9606</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC2|NAT-2|PNAT</td>\n",
       "      <td>MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156...</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 2|N-acetyltransf...</td>\n",
       "      <td>20220703</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        #tax_id Symbol LocusTag                   Synonyms  \\\n",
       "GeneID                                                       \n",
       "1          9606   A1BG        -       A1B|ABG|GAB|HYST2477   \n",
       "2          9606    A2M        -  A2MD|CPAMD5|FWP007|S863-7   \n",
       "3          9606  A2MP1        -                       A2MP   \n",
       "9          9606   NAT1        -       AAC1|MNAT|NAT-1|NATI   \n",
       "10         9606   NAT2        -            AAC2|NAT-2|PNAT   \n",
       "\n",
       "                                                  dbXrefs chromosome  \\\n",
       "GeneID                                                                 \n",
       "1       MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410...         19   \n",
       "2       MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899...         12   \n",
       "3       HGNC:HGNC:8|Ensembl:ENSG00000256069|AllianceGe...         12   \n",
       "9       MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171...          8   \n",
       "10      MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156...          8   \n",
       "\n",
       "       map_location                         description    type_of_gene  \\\n",
       "GeneID                                                                    \n",
       "1          19q13.43              alpha-1-B glycoprotein  protein-coding   \n",
       "2          12p13.31               alpha-2-macroglobulin  protein-coding   \n",
       "3          12p13.31  alpha-2-macroglobulin pseudogene 1          pseudo   \n",
       "9              8p22               N-acetyltransferase 1  protein-coding   \n",
       "10             8p22               N-acetyltransferase 2  protein-coding   \n",
       "\n",
       "       Symbol_from_nomenclature_authority  \\\n",
       "GeneID                                      \n",
       "1                                    A1BG   \n",
       "2                                     A2M   \n",
       "3                                   A2MP1   \n",
       "9                                    NAT1   \n",
       "10                                   NAT2   \n",
       "\n",
       "       Full_name_from_nomenclature_authority Nomenclature_status  \\\n",
       "GeneID                                                             \n",
       "1                     alpha-1-B glycoprotein                   O   \n",
       "2                      alpha-2-macroglobulin                   O   \n",
       "3         alpha-2-macroglobulin pseudogene 1                   O   \n",
       "9                      N-acetyltransferase 1                   O   \n",
       "10                     N-acetyltransferase 2                   O   \n",
       "\n",
       "                                       Other_designations  Modification_date  \\\n",
       "GeneID                                                                         \n",
       "1       alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...           20220703   \n",
       "2       alpha-2-macroglobulin|C3 and PZP-like alpha-2-...           20220703   \n",
       "3                       pregnancy-zone protein pseudogene           20220513   \n",
       "9       arylamine N-acetyltransferase 1|N-acetyltransf...           20220522   \n",
       "10      arylamine N-acetyltransferase 2|N-acetyltransf...           20220703   \n",
       "\n",
       "       Feature_type  \n",
       "GeneID               \n",
       "1                 -  \n",
       "2                 -  \n",
       "3                 -  \n",
       "9                 -  \n",
       "10                -  "
      ]
     },
     "execution_count": 189,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ncbi.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|| 26415/26415 [01:23<00:00, 317.11it/s] "
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8422\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "nodes = {} \n",
    "edges = {}\n",
    "no_data = set()\n",
    "gene_types = {}\n",
    "for gene in tqdm(human_correlation_rank.index):\n",
    "    # check if we want this gene\n",
    "    gene_symbol, gene_id = get_info(gene)\n",
    "    if gene_symbol and gene_id:\n",
    "        if gene_id not in nodes:\n",
    "            synonyms = list(set(ncbi.at[int(gene_id), \"Synonyms\"].split(\"|\")))\n",
    "            XRefs = {}\n",
    "            gene_type = ncbi.at[int(gene_id), \"type_of_gene\"]\n",
    "            if gene_type not in gene_types:\n",
    "                gene_types[gene_type] = 0\n",
    "            gene_types[gene_type] += 1\n",
    "            for i in ncbi.at[int(gene_id), \"dbXrefs\"].split(\"|\"):\n",
    "                split = i.split(\":\")\n",
    "                k = split[0]\n",
    "                v = \":\".join(split[1:])\n",
    "                XRefs[k] = v\n",
    "            scores = {}\n",
    "            if gene_symbol in gene_scores.index:\n",
    "                for k,v in gene_scores.loc[gene_symbol].items():\n",
    "                    scores[k.replace(\".\", \"_\")] = v\n",
    "            \n",
    "            nodes[gene_id] = {\n",
    "                \"type\": \"Gene\",\n",
    "                \"properties\": {\n",
    "                    \"id\": gene_id,\n",
    "                    \"label\": gene_symbol,\n",
    "                    \"uri\": \"https://www.ncbi.nlm.nih.gov/gene/%s\"%gene_id,\n",
    "                    **scores\n",
    "                }\n",
    "            }\n",
    "#             if len(synonyms):\n",
    "#                 nodes[gene_id][\"properties\"][\"synonyms\"] = synonyms\n",
    "#             if len(XRefs):\n",
    "#                 nodes[gene_id][\"properties\"][\"XRefs\"] = [\"%s: %s\"%(k,v) for k,v in XRefs.items()]\n",
    "        rank = 0\n",
    "        count = 0\n",
    "        sorted_genes = human_correlation_rank[gene].sort_values().index\n",
    "        while count < 5:\n",
    "            coexpressed_gene = sorted_genes[rank]\n",
    "            coexpressed_symbol, coexpressed_id = get_info(coexpressed_gene)\n",
    "            if coexpressed_symbol and coexpressed_id:\n",
    "                if coexpressed_id not in nodes:\n",
    "                    gene_type = ncbi.at[int(coexpressed_id), \"type_of_gene\"]\n",
    "                    if gene_type not in gene_types:\n",
    "                        gene_types[gene_type] = 0\n",
    "                    gene_types[gene_type] += 1\n",
    "                    synonyms = list(set(ncbi.at[int(coexpressed_id), \"Synonyms\"].split(\"|\")))\n",
    "                    XRefs = {}\n",
    "                    for i in ncbi.at[int(coexpressed_id), \"dbXrefs\"].split(\"|\"):\n",
    "                        split = i.split(\":\")\n",
    "                        k = split[0]\n",
    "                        v = \":\".join(split[1:])\n",
    "                        XRefs[k] = v\n",
    "                    scores = {}\n",
    "                    if coexpressed_symbol in gene_scores.index:\n",
    "                        for k,v in gene_scores.loc[coexpressed_symbol].items():\n",
    "                            scores[k.replace(\".\", \"_\")] = v\n",
    "                    nodes[coexpressed_id] = {\n",
    "                        \"type\": \"Gene\",\n",
    "                        \"properties\": {\n",
    "                            \"id\": coexpressed_id,\n",
    "                            \"label\": coexpressed_symbol,\n",
    "                            \"uri\": \"https://www.ncbi.nlm.nih.gov/gene/%s\"%coexpressed_id,\n",
    "                            **scores\n",
    "                        }\n",
    "                    }\n",
    "#                     if len(synonyms):\n",
    "#                         nodes[coexpressed_id][\"properties\"][\"synonyms\"] = synonyms\n",
    "#                     if len(XRefs):\n",
    "#                         nodes[coexpressed_id][\"properties\"][\"XRefs\"] = [\"%s: %s\"%(k,v) for k,v in XRefs.items()]\n",
    "                    \n",
    "                edge_1 = (gene_id, coexpressed_id)\n",
    "                edge_2 = (coexpressed_id, gene_id)\n",
    "                if edge_1 not in edges and edge_2 not in edges:\n",
    "                    edges[edge_1] = {\n",
    "                        \"type\": \"Relation\",\n",
    "                        \"source\": gene_id,\n",
    "                        \"relation\": \"ARCHS4 (negatively correlated)\",\n",
    "                        \"target\": coexpressed_id,\n",
    "                        \"properties\": {\n",
    "                            \"resource\": \"ARCHS4\",\n",
    "                            \"source_label\": gene_symbol,\n",
    "                            \"target_label\": coexpressed_symbol,\n",
    "                            \"label\": \"negatively correlated\",\n",
    "                            \"id\": \"%s is negatively correlated with %s\"%(gene_symbol, coexpressed_symbol),\n",
    "                            \"correlation_coefficient\": human_correlation.at[coexpressed_gene, gene]\n",
    "                        }\n",
    "                    }\n",
    "                count += 1\n",
    "            rank += 1\n",
    "            \n",
    "        count = 0\n",
    "        rank = len(sorted_genes) - 1\n",
    "        \n",
    "        while count < 5:\n",
    "            coexpressed_gene = sorted_genes[rank]\n",
    "            if not gene == coexpressed_gene:\n",
    "                coexpressed_symbol, coexpressed_id = get_info(coexpressed_gene)\n",
    "                if coexpressed_symbol and coexpressed_id:\n",
    "                    if coexpressed_id not in nodes:\n",
    "                        nodes[coexpressed_id] = {\n",
    "                            \"type\": \"Gene\",\n",
    "                            \"properties\": {\n",
    "                                \"id\": coexpressed_id,\n",
    "                                \"label\": coexpressed_symbol\n",
    "                            }\n",
    "                        }\n",
    "                    edge_1 = (gene_id, coexpressed_id)\n",
    "                    edge_2 = (coexpressed_id, gene_id)\n",
    "                    if edge_1 not in edges and edge_2 not in edges:\n",
    "                        edges[edge_1] = {\n",
    "                            \"type\": \"Relation\",\n",
    "                            \"source\": gene_id,\n",
    "                            \"relation\": \"ARCHS4 (positively correlated)\",\n",
    "                            \"target\": coexpressed_id,\n",
    "                            \"properties\": {\n",
    "                                \"resource\": \"ARCHS4\",\n",
    "                                \"source_label\": gene_symbol,\n",
    "                                \"target_label\": coexpressed_symbol,\n",
    "                                \"label\": \"positively correlated\",\n",
    "                                \"id\": \"%s is positively correlated with %s\"%(gene_symbol, coexpressed_symbol),\n",
    "                                \"correlation_coefficient\": human_correlation.at[coexpressed_gene, gene]\n",
    "                            }\n",
    "                        }\n",
    "                    count += 1\n",
    "            rank -= 1\n",
    "    else:\n",
    "        no_data.add(gene)\n",
    "print(len(no_data))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'protein-coding': 10230}"
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17964, 170801)"
      ]
     },
     "execution_count": 192,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(nodes), len(edges)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 193,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|| 17964/17964 [00:00<00:00, 62715.56it/s]\n",
      "100%|| 170801/170801 [00:06<00:00, 24436.40it/s]\n"
     ]
    }
   ],
   "source": [
    "serialization_v1 = {\n",
    "    \"version\": \"1\",\n",
    "    \"nodes\": nodes,\n",
    "    \"edges\": list(edges.values())\n",
    "}\n",
    "with open(\"results/archs4_coexpression.v1.json\", \"w\") as o:\n",
    "    o.write(json.dumps(serialization_v1, indent=2))\n",
    "\n",
    "nodes_v2 = {}\n",
    "for k,v in tqdm(nodes.items()):\n",
    "    nodes_v2[k] = typer(v)\n",
    "edges_v2 = []\n",
    "for i in tqdm(edges.values()):\n",
    "    edges_v2.append(\n",
    "        typer(i)\n",
    "    )\n",
    "\n",
    "serialization_v2 = {\n",
    "    \"version\": \"2\",\n",
    "    \"nodes\": nodes_v2,\n",
    "    \"edges\": edges_v2\n",
    "}\n",
    "with open(\"results/archs4_coexpression.v2.json\", \"w\") as o:\n",
    "    o.write(json.dumps(serialization_v2, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### copy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 237,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in glob.glob('results/*.v1.json'):\n",
    "    with open(i) as o:\n",
    "        new_file = i.replace(\"v1\", \"valid\").replace('results', 'ingestion')\n",
    "        with open(new_file, 'w') as w:\n",
    "            w.write(o.read())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "file: results/reprotox_serialization.v1.json, nodes: 1427, edges: 2252\n",
      "file: results/drugsto_faers_female.v1.json, nodes: 126, edges: 193\n",
      "file: results/sigcom_lincs_drug_similarity.v1.json, nodes: 4523, edges: 20785\n",
      "file: results/drugsto_faers_male.v1.json, nodes: 126, edges: 193\n",
      "file: results/hpo.v1.json, nodes: 5195, edges: 126862\n",
      "file: results/sigcom_lincs_serialization.v1.json, nodes: 8942, edges: 225509\n",
      "file: results/archs4_coexpression.v1.json, nodes: 17964, edges: 170801\n",
      "file: results/idg_drug_targets.v1.json, nodes: 2394, edges: 7324\n"
     ]
    }
   ],
   "source": [
    "for i in glob.glob('results/*.v1.json'):\n",
    "    with open(i) as o:\n",
    "        serialization = json.loads(o.read())\n",
    "        print(\"file: %s, nodes: %s, edges: %s\"%(i, len(serialization[\"nodes\"]),len(serialization[\"edges\"])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.13 ('venv': venv)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "vscode": {
   "interpreter": {
    "hash": "43fa70dbc3e0e900c09e1e9e147b0eb772ee319912638d9a00663e86ab5de451"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
